feat: adds first cut at null-label post (#14)
masterpointio/masterpoint.io: content/updates/terraform-null-label.md

Our good colleagues at [Cloud Posse](https://cloudposse.com/) have come up with a much more elegant solution (and huge shout out to them for doing so). They have developed a module called [terraform-null-label](https://github.com/cloudposse/terraform-null-label), which offers a much cleaner approach to this problem. This module generates and standardizes names and tags for resources across an entire Terraform environment, ensuring consistency. As part of this post, we’d like to take some time to explore the cleanest solution to this problem we’ve ever seen.

## The Importance of Consistent Naming

In the early stages of a startup (or any project stack), the infrastructure is often simple and straightforward. The server and database are probably named something like `server01` and `database01`, respectively. The configuration files are easily accessible, and the team, being small, is intimately familiar with the system's architecture. The naming conventions, at this point, are uncomplicated and intuitive.

> However, as the project grows and evolves, so does its infrastructure. The once simple system expands to accommodate multiple customer-facing applications and a dozen other resources enter the mix. Infrastructure as Code (IaC) is implemented to manage and provision the increasingly complex tech stack. Amidst this growth, `database01` persists, but what about `database10`?. It's now `database10-us-east1-prod`; absolutely more descriptive and useful, but no one wants to rename `database01 `because it’s probably hardcoded into more than a few places.

Reviewer: ```suggestion
However, as the project grows and evolves, so does its infrastructure. The once simple system expands to accommodate multiple customer-facing applications and a dozen other resources enter the mix. Infrastructure as Code (IaC) is implemented to manage and provision the increasingly complex tech stack. Amidst this growth, `database01` persists, but what about `database10`? It's now `database10-us-east1-prod`; absolutely more descriptive and useful, but no one wants to rename `database01` because it’s probably hardcoded into more than a few places.
```

The issue of inconsistent naming becomes even more pronounced when we consider tags. What was once simply "backend" now has new variations like "backend-auth" and "backend-job-processor". Are those cloud resources getting tagged correctly so whoever is paying the bill knows that the "job processor" is costing 10x what the "auth" service is? Probably not.

As the stack expands and you need to hire more engineers, the cognitive load and spool-up time for each new hire grows and grows. As it becomes harder to understand and categorize and "know about" the infrastructure, it becomes a bigger security/attack surface. Eventually billing and resource attribution becomes a problem too.

> ## The Fix: terraform-null-label

Reviewer: I feel like `The Solution` sounds more solid for the problem described about. Unless there is some connotation/reference I'm missing here.

Reviewer 2: I think "The Solution" might be better too.

Author: ```suggestion
## The Solution
```

> Well, the **real** fix is consistent naming and tagging, but the challenge has always been getting a growing, changing roster of engineers to decide on, implement, and enforce a standard _over time_.

Reviewer 2: ```suggestion
Let's face it: naming things is hard. We want consistent naming and tagging, but the challenge has always been getting a growing, changing roster of engineers to decide on, implement, and enforce a consistent standard _over time_.
```

Good luck with that. Maybe you’ve written down your standards in documentation, but that’s hard to find, rarely read, and not easily enforceable.

> Why not use code?

Reviewer: This looks a bit lost in the text. Should be a next-level heading? Or maybe just continue the next sentence in the same line.

> Why not use code?
> 
> The `terraform-null-label` module, available at [https://github.com/cloudposse/terraform-null-label](https://github.com/cloudposse/terraform-null-label), offers a much better solution. We’ll look at some example usage in the next section but at a high level It allows users to define up to six primary inputs, referred to as "ID elements". These include:

Author: ```suggestion
Why not use code? The `terraform-null-label` module, available at [https://github.com/cloudposse/terraform-null-label](https://github.com/cloudposse/terraform-null-label), offers a much better solution. We’ll look at some example usage in the next section but at a high level It allows users to define up to six primary inputs, referred to as "ID elements". These include:
```

1. **Namespace:** A short abbreviation of the company name, typically 2-4 letters.

> 2. **Tenant:** This is used for resources dedicated to a specific customer or team. (For a team like "data analytics" => "da")

Reviewer: ```suggestion
2. **Tenant:** This is used for resources dedicated to a specific customer or team (for a team like "data analytics" => "da").
```

> 3. **Environment:** This refers to the AWS region. (Note that this nomenclature may be inverse to what most engineers are used to).

Reviewer: ```suggestion
3. **Environment:** This refers to the AWS region (note that this nomenclature may be inverse to what most engineers are used to).
```

Reviewer 2: Just for a bit of history, no need to add it. 

Technically, this is how Cloud Posse has been using `environment` for our purposes. We struggled on what to use or adding another parameter like `region`. We didn't want to use region because the canonical region (e.g. `us-east-1`) is frequently needed in modules but is not the "region" used for naming/labeling purposes due to length constraints.

I don't see any reason why others cannot use `environment` to mean what we call stage. 

Of course, in `atmos` some of these are more opinionated, but that's something we intend to eliminate in Atmos soon, so arbitrary contexts can be used.

4. **Stage:** This indicates the stage of the application, such as `dev`, `stage`, `prod`. This is what we would normally call an `environment` in most settings, but legacy naming within the module has made this a bit funky.

> 5. **Name:** This is the name of the component or service, such as 'rds','eks', or `backend-auth`.

Reviewer: ```suggestion
5. **Name:** This is the name of the component or service, such as `rds`, `eks`, or `backend-auth`.
```

> 6. **Attributes:** This is an arbitrary list of strings that can be used to add additional descriptors to the final ID.

Reviewer 2: This is correct. The term we often use is to help "disambiguate" a resource.

The `terraform-null-label` module provides a standardized set of outputs that combine these various ID elements for easy usage across resources and modules. For example, if the following ID elements were supplied to a label module named `frontend_label`:

```hcl

>   namespace  = "mp" # Shorthand for Masterpoint
>   environment = "uw2" # Shorthand for the AWS us-west-2 region
>   stage      = "prod"
>   name       = "frontend"
>   attributes = ["public"]

Reviewer: ```suggestion
```suggestion
  namespace   = "mp" # Shorthand for Masterpoint
  environment = "uw2" # Shorthand for the AWS us-west-2 region
  stage       = "prod"
  name        = "frontend"
  attributes  = ["public"]
```

These output values are now available to apply to any "name" or "tags" input for any Terraform resource or module, saving engineers the cognitive load of needing to remember and apply the correct naming or tagging structure.

Using `terraform-null-label` for consistency in naming and tagging is especially important for billing and any FinOps processes. One of the biggest challenges in managing cloud costs is properly attributing resource usage across different teams and business units; without consistency in tagging, it’s a nearly impossible task to do effectively.

## Utilizing the `terraform-null-label` module

Below, we have a hypothetical example infrastructure that will be a simple implementation of an AWS Load Balancer. Most platform engineers have encountered these resources before (and their corresponding IaC), and are likely aware of the massive waste of time and cognitive effort it is to remember to name and tag every resource consistently when more resources continue to get added.

> Enter `terraform-null-label`.

Reviewer: Similar as above, this looks a bit "lost" to me.

> Enter `terraform-null-label`.
> 
> Here’s an example infrastructure, with `terraform-null-label` applied:

Author: ```suggestion
Enter `terraform-null-label`. Here’s an example infrastructure, with the label module applied:
```

```hcl
  module "public_alb_label" {
    source  = "cloudposse/label/null"
    version = "0.25.0"

    namespace   = "mp"
    stage       = "prod"
    environment = "uw2"
    name        = "alb"
    attributes  = ["public"]

    tags = {
      BusinessUnit       = "ENG"
      Team               = "Reporting"
      ManagedByTerraform = "True"
    }
  }

  # ALB
  resource "aws_lb" "default" {

    # ID is the concatenation result of all ID element inputs i.e. `mp-uw2-prod-alb-public`
    name               = module.public_alb_label.id

    internal           = false
    load_balancer_type = "application"
    security_groups    = [aws_security_group.alb_sg.id]
    subnets            = ["subnet-abc12345", "subnet-xyz67890"]

    # terraform-null-label outputs the merged tags input and all ID element inputs as a map value
    # i.e. { namespace: "mp", stage: "prod", Team: "Reporting", ManagedByTerraform: "True", ... }

>     tags               = module.public_alb_label.tags

Reviewer: ```suggestion
    tags = module.public_alb_label.tags
```

  }

  resource "aws_security_group" "alb_sg" {
    name        = module.public_alb_label.id
    description = "Allow inbound traffic over HTTPS"
    vpc_id      = "vpc-12345678"
    tags        = module.public_alb_label.tags

    ingress {
      from_port   = 443
      to_port     = 443
      protocol    = "tcp"
      cidr_blocks = ["0.0.0.0/0"]
    }
  }
```

It’s worth noting that in our examples, we’re using the AWS provider, but the label module can be used anywhere in Terraform regardless of provider or use-case.

In the above example, the `terraform-null-label` module lets engineers define some common-sense values for the ID elements like stage and namespace, and then it can be used to apply a consistent name and tagging scheme to every resource and/or module that the team uses. "Write once, read many" and DRY (Don’t repeat yourself) are important factors we talk a lot about within good application code, and null-label enables us to easily incorporate those ideas into our platform code.

> Now, obviously that’s a very simplified and abstract example. In complex configurations, there are likely going to be many resources and modules with different usage, and in this case even with `terraform-null-label` it can start to get cumbersome to have to define new label modules each time you need to have variations in your ID elements. There is a solution to that however as one of the module’s most powerful features addresses exactly that: the `context` input + output value.

Reviewer 2: It may be worth mentioning that this is just taking a common behavior from languages like Perl or JavaScript, that typically pass around something called a "context." 

Just for clarification, and this doesn't need to be included in the article: In older JavaScript patterns, particularly before the ES6 arrow functions, the this context is sometimes explicitly passed as the first argument using methods JavaScript. In Perl, context is often conveyed through a `self` variable passed to methods. The context provides information that helps a function or method operate in a particular scope or environment.

e.g. 
```package Person;

sub new {
    my $class = shift;
    my $self = {
        name => shift,
    };
    bless $self, $class;
    return $self;
}

sub say_hello {
    my ($self) = @_;
    print "Hello, my name is $self->{name}\n";
}

# Main program
my $person = Person->new("John");
$person->say_hello();  # Output: "Hello, my name is John"
```

e.g. in JavaScript
```
function sayHello(greeting) {
    console.log(`${greeting}, my name is ${this.name}`);
}

const obj = {
    name: "John"
};

sayHello.call(obj, "Hello");  // Output: "Hello, my name is John"
```

Author: Ah great callout @osterman, I've seen the same pattern in other languages as well, but I never made the connection to null-label context. 

We found this post was getting a bit long, so we're going to follow up with a whole post on `context.tf` (and maybe mixins in general). When we do, we'll be sure to include this as I'm sure it'll help people get what is going on here!

      Team               = "Reporting"
      ManagedByTerraform = "True"
    }
  }

  module "private_alb_label" {
    source  = "cloudposse/label/null"
    version = "0.25.0"

    attributes  = ["private"]

    # Here's the important bit
    context = module.public_alb_label.context
  }
```

By passing in the context output from the first label module, all of the original values are present, with only the defined ones overridden. With that, you get:

* `module.public_alb_label.id` equates to `mp-uw2-prod-alb-public`
* `module.public_alb_label.tags` equates to:

   ```yaml
   Namespace: "mp"
   Environment: "uw2"
   Stage: "prod"
   Name: "alb"
   Attributes: "public"
   BusinessUnit: "ENG"
   Team: "Reporting"
   ManagedByTerraform: "True"

* `module.private_alb_label.id` equates to `mp-uw2-prod-alb-private`
* `module.private_alb_label.tags` equates to:

   ```yaml
   Namespace: "mp"
   Environment: "uw2"
   Stage: "prod"
   Name: "alb"
   Attributes: "private"
   BusinessUnit: "ENG"
   Team: "Reporting"
   ManagedByTerraform: "True"

The new label module only needs to have new values defined; the previous values are still present and do not need to be redefined.

> Admittedly, that’s still a pretty simplified example. Real Terraform configurations often employ root and child modules and various workspaces / folders for each; to this point our examples have been constrained to one workspace, working directory, and statefile. There is a nice solution to simplify that process however via the `context.tf` file, but we’re going to explore that advanced usage in a separate article, so be sure to watch this space!

Reviewer: It's probably worth to add a link to https://github.com/cloudposse/terraform-null-label/blob/main/exports/context.tf file for some context (hey, it's a wordplay!). At least to provide some understanding that it's a file that lives within this module for those who a new to it.

The issue of inconsistent naming becomes even more pronounced when we consider tags. What was once simply "backend" now has new variations like "backend-auth" and "backend-job-processor". Are those cloud resources getting tagged correctly so whoever is paying the bill knows that the "job processor" is costing 10x what the "auth" service is? Probably not.

As the stack expands and you need to hire more engineers, the cognitive load and spool-up time for each new hire grows and grows. As it becomes harder to understand and categorize and "know about" the infrastructure, it becomes a bigger security/attack surface. Eventually billing and resource attribution becomes a problem too.

## The Fix: terraform-null-label

Well, the **real** fix is consistent naming and tagging, but the challenge has always been getting a growing, changing roster of engineers to decide on, implement, and enforce a standard _over time_.

Good luck with that. Maybe you’ve written down your standards in documentation, but that’s hard to find, rarely read, and not easily enforceable.

Why not use code?

> The `terraform-null-label`` module, available at [https://github.com/cloudposse/terraform-null-label](https://github.com/cloudposse/terraform-null-label), offers a much better solution. We’ll look at some example usage in the next section but at a high level It allows users to define up to six primary inputs, referred to as "ID elements". These include:

Reviewer 3: There's an extra backtick here that's throwing off the text formatting:
> The `terraform-null-label`` module

Author: Good callout -- Thanks Nate! Fixing below.

> The `terraform-null-label`` module, available at [https://github.com/cloudposse/terraform-null-label](https://github.com/cloudposse/terraform-null-label), offers a much better solution. We’ll look at some example usage in the next section but at a high level It allows users to define up to six primary inputs, referred to as "ID elements". These include:

Author: ```suggestion
The `terraform-null-label` module, available at [https://github.com/cloudposse/terraform-null-label](https://github.com/cloudposse/terraform-null-label), offers a much better solution. We’ll look at some example usage in the next section but at a high level It allows users to define up to six primary inputs, referred to as "ID elements". These include:
```

Going back to our example, what if we needed to deploy another internal load balancer alongside the public one? That is going to share a lot of the same label id elements with a couple of overrides. Here’s how easy that is:

```hcl
  module "public_alb_label" {
    source  = "cloudposse/label/null"
    version = "0.25.0"

    namespace   = "mp"
    stage       = "prod"
    environment = "uw2"
    name        = "alb"
    attributes  = ["public"]

    tags = {
      BusinessUnit       = "ENG"
      Team               = "Reporting"
      ManagedByTerraform = "True"
    }
  }

  module "private_alb_label" {
    source  = "cloudposse/label/null"
    version = "0.25.0"

    attributes  = ["private"]

    # Here's the important bit
    context = module.public_alb_label.context
  }
```

By passing in the context output from the first label module, all of the original values are present, with only the defined ones overridden. With that, you get:

* `module.public_alb_label.id` equates to `mp-uw2-prod-alb-public`
* `module.public_alb_label.tags` equates to:

   ```json
   {

>      Namespace: "mp",

Reviewer 3: These blobs of data aren't valid JSON, so the syntax highlighter is showing this weirdly in the rendered page. What language was this meant to be?

Author: I dunno, I was trying to show them as JSON. I'll update it to be less JSON-y (JSON sucks).

>    ```json
>    {
>      Namespace: "mp",
>      Environment: "uw2",
>      Stage: "prod",
>      Name: "alb",
>      Attributes: "public",
>      BusinessUnit: "ENG",
>      Team: "Reporting",
>      ManagedByTerraform: "True",
>    }
>    ```

Author: ```suggestion
   ```yaml
   Namespace: "mp"
   Environment: "uw2"
   Stage: "prod"
   Name: "alb"
   Attributes: "public"
   BusinessUnit: "ENG"
   Team: "Reporting"
   ManagedByTerraform: "True"
```
```

* `module.private_alb_label.id` equates to `mp-uw2-prod-alb-private`
* `module.private_alb_label.tags` equates to:

>    ```json
>    {
>      Namespace: "mp",
>      Environment: "uw2",
>      Stage: "prod",
>      Name: "alb",
>      Attributes: "private",
>      BusinessUnit: "ENG",
>      Team: "Reporting",
>      ManagedByTerraform: "True",
>    }
>    ```

Author: ```suggestion
   ```yaml
   Namespace: "mp"
   Environment: "uw2"
   Stage: "prod"
   Name: "alb"
   Attributes: "private"
   BusinessUnit: "ENG"
   Team: "Reporting"
   ManagedByTerraform: "True"
   ```
```
